polar-core/src/bindings.rs

```diff
@@ -0,0 +1,313 @@
+                VariableState::Bound(value) => value,
+                _ => term.clone(),
+            },
+            Value::RestVariable(v) => match self.variable_state(v) {
+                VariableState::Bound(value) => match value.value() {
+                    Value::List(l) if has_rest_var(l) => self.deref(&value),
+                    _ => value,
+                },
+                _ => term.clone(),
+            },
+            _ => term.clone(),
+        }
+    }
+
+    /// Dereference all variables in term, including within nested structures like
+    /// lists and dictionaries.
+    /// Do not dereference variables inside expressions.
+    pub fn deep_deref(&self, term: &Term) -> Term {
+        pub struct Derefer<'a> {
+            binding_manager: &'a BindingManager,
+        }
+
+        impl<'a> Derefer<'a> {
+            pub fn new(binding_manager: &'a BindingManager) -> Self {
+                Self { binding_manager }
+            }
+        }
+
+        impl<'a> Folder for Derefer<'a> {
+            fn fold_term(&mut self, t: Term) -> Term {
+                match t.value() {
+                    Value::List(_) => fold_term(self.binding_manager.deref(&t), self),
+                    Value::Variable(_) | Value::RestVariable(_) => {
+                        let derefed = self.binding_manager.deref(&t);
+                        match derefed.value() {
+                            Value::Expression(_) => t,
+                            _ => fold_term(derefed, self),
+                        }
+                    }
+                    _ => fold_term(t, self),
+                }
+            }
+        }
+
+        Derefer::new(self).fold_term(term.clone())
+    }
+
+    /// Check the state of `variable`.
+    pub fn variable_state(&self, variable: &Symbol) -> VariableState {
+        self.variable_state_at_point(variable, self.bsp())
+    }
+
+    // TODO: Get rid of this, only used in inverter.
+    /// Check the state of `variable` at `bsp`.
+    pub fn variable_state_at_point(&self, variable: &Symbol, bsp: Bsp) -> VariableState {
+        let mut path = vec![variable];
+        while let Some(value) = self.value(path.last().unwrap(), bsp) {
+            match value.value() {
+                Value::Expression(e) => return VariableState::Partial(e.clone()),
+                Value::Variable(v) | Value::RestVariable(v) => {
+                    if v == variable {
+                        return VariableState::Cycle(path.into_iter().cloned().collect());
+                    } else {
+                        path.push(v);
+                    }
+                }
+                _ => return VariableState::Bound(value.clone()),
+            }
+        }
+        VariableState::Unbound
+    }
+
+    /// Add `term` as a constraint.
+    pub fn add_constraint(&mut self, term: &Term) -> PolarResult<()> {
+        let Operation { operator: op, args } = term.value().as_expression().unwrap();
+        assert!(
```

Comment:
Not for this PR, but just occurred to me that using assert here is not a great idea. We could have our own macro to return some PolarError

Reply:
Or maybe debug_assert plus return error in release